backends/gstreamer/src/player.rs

```diff
@@ -324,32 +364,97 @@ impl GStreamerPlayer {
+                    if let Some(el) = msg.get_src().map(|s| s.downcast::<gst::Element>().unwrap()) {
+                        use gst_gl::ContextGLExt;
+
+                        let context_type = ctxt.get_context_type();
+                        if context_type == *gst_gl::GL_DISPLAY_CONTEXT_TYPE {
+                            let context = gst::Context::new(context_type, true);
+                            context.set_gl_display(&gl_display);
+                            el.set_context(&context);
+                        } else if context_type == "gst.gl.app_context" {
+                            let mut context = gst::Context::new(context_type, true);
+                            {
+                                let mut s = context.get_mut().unwrap().get_mut_structure();
+                                s.set_value("context", gl_context.to_send_value());
+                            }
+                            el.set_context(&context);
+                        }
+                    }
+                }
+                _ => (),
+            }
+
+            gst::BusSyncReply::Pass
+        });
+
+        Ok(())
+    }
+
     fn setup_video_sink(
         &self,
         player: &gst_player::Player,
     ) -> Result<gst_app::AppSink, PlayerError> {
         let pipeline = player.get_pipeline();
 
-        let video_sink = gst::ElementFactory::make("appsink", None)
-            .ok_or(PlayerError::Backend("appsink creation failed".to_owned()))?;
-
-        pipeline
-            .set_property("video-sink", &video_sink.to_value())
-            .map_err(|e| PlayerError::Backend(e.to_string()))?;
-
-        let video_sink = video_sink.dynamic_cast::<gst_app::AppSink>().unwrap();
-        video_sink.set_caps(&gst::Caps::new_simple(
-            "video/x-raw",
-            &[
-                ("format", &"BGRA"),
-                ("pixel-aspect-ratio", &gst::Fraction::from((1, 1))),
-            ],
-        ));
+        let appsink = gst::ElementFactory::make("appsink", None)
+            .ok_or(PlayerError::Backend("appsink creation failed".to_owned()))?
+            .dynamic_cast::<gst_app::AppSink>()
+            .unwrap();
+
+        if self.gl_context.borrow().is_some() {
+            let glsink = gst::ElementFactory::make("glsinkbin", None)
+                .ok_or(PlayerError::Backend("glsinkbin creation failed".to_owned()))?;
+
+            let caps = gst::Caps::builder("video/x-raw")
+                .features(&[&gst_gl::CAPS_FEATURE_MEMORY_GL_MEMORY])
+                .field("format", &gst_video::VideoFormat::Rgba.to_string())
+                .field("texture-target", &"2D")
+                .build();
+            appsink.set_caps(&caps);
+
+            glsink
+                .set_property("sink", &appsink)
+                .map_err(|e| PlayerError::Backend(e.to_string()))?;
+
+            pipeline
+                .set_property("video-sink", &glsink.to_value())
+                .map_err(|e| PlayerError::Backend(e.to_string()))?;
```

Comment:
If any of these `set_property()` fail you have a programming error. You probably want to make that an `expect()` instead.